src/CommandQoSController.cpp

```diff
@@ -0,0 +1,133 @@
+#include "CommandQoSController.hpp"
+#include "CommandRunner.hpp"
+#include "Helpers.hpp"
+
+#include <google/protobuf/util/json_util.h>
+
+#include <process/defer.hpp>
+#include <process/dispatch.hpp>
+
+using ::mesos::modules::Module;
+using ::mesos::ResourceUsage;
+using ::mesos::slave::QoSCorrection;
+
+using std::list;
+using std::string;
+
+using process::Future;
+using process::Failure;
+
+namespace criteo {
+
+namespace mesos {
+
+class CommandQoSControllerProcess
+    : public process::Process<CommandQoSControllerProcess> {
+ public:
+  CommandQoSControllerProcess(
+      const lambda::function<Future<ResourceUsage>()>& usage,
+      const Option<Command>& correctionsCommand, bool isDebugMode);
+
+  virtual process::Future<list<::mesos::slave::QoSCorrection>> corrections();
+  virtual list<::mesos::slave::QoSCorrection> m_corrections(
+      const ResourceUsage& m_usage);
+  virtual inline const Option<Command>& correctionsCommand() const {
+    return m_correctionsCommand;
+  }
+
+ protected:
+  const lambda::function<Future<ResourceUsage>()> m_usage;
+
+ private:
+  const Option<Command> m_correctionsCommand;
+  bool m_isDebugMode;
+};
+
+CommandQoSControllerProcess::CommandQoSControllerProcess(
+    const lambda::function<process::Future<ResourceUsage>()>& usage,
+    const Option<Command>& correctionsCommand, bool isDebugMode)
+    : m_usage(usage),
+      m_correctionsCommand(correctionsCommand),
+      m_isDebugMode(isDebugMode) {}
+
+list<::mesos::slave::QoSCorrection> CommandQoSControllerProcess::m_corrections(
+    const ResourceUsage& m_usage) {
+  LOG(INFO) << "Start Corrections timer_start_module QoSController";
+  list<QoSCorrection> listCorrection;
+  if (m_correctionsCommand.isNone()) {
+    return list<QoSCorrection>();
+    ;
+  }
+  logging::Metadata metadata = {"0", "corrections"};
+
+  CommandRunner cmdRunner = CommandRunner(m_isDebugMode, metadata);
+
+  string input;
+  google::protobuf::util::MessageToJsonString(m_usage, &input);
+  LOG(INFO) << "Start QoSscript timer_start_script QoSController";
+  Try<string> output = cmdRunner.run(m_correctionsCommand.get(), input);
+  LOG(INFO) << "End QoSscript timer_end_script QoSController";
+
+  if (output.isError()) {
+    LOG(INFO) << output.error();
+    return list<QoSCorrection>();
+  }
+  // turn the json into QoSCorrections
+  // need to turn a string into list of message, going to use split
+  // actual output = {QoSCorrection}; {QoSCorrection};...
+  // wished output = [{QoSCorrection}, {QoSCorrection}, ...]
+
+  std::string stroutput = output.get();
+  std::string delimiter = ";";
```

Comment:
I'm not sure to get why the script could not return the correct format (which seems to be map nicely with json)

Reply:
Actually the problem come from the way message are supported.
Since the return must be an **array of QoSCorrection** there is noway to call for any message converting like **JsonStringtoMessage()**.
I tried to convert the **JsonArray to Array** in c++ but did not find and easy way to do it, a part from calling external library.

Each **QoSCorrection** look like this:
{"type":"KILL","kill":{"executorId":{"value":"app"},"frameworkId"{"value":"frameworkID"},"containerId":{"value":"containerID"}}}
Which kind of complicated the the way to split it so i use a special character as separator

If you have any good way to do i would gladly take it!

Reply:
I agree with Greg, this part looks a bit fragile. I've read your answer but have somle additional questions:
* what is that "QoSscript timer_start_script QoSController" script? If this is something on our side, should it not return directly a json, like all other commands run?
* if it's not in our hands, maybe there should be an intermediate wrapper to do it anyway, because I don't feel this kind of things needs to be done in this module.

Reply:
thanks for the feeback, the "QoSscript timer_start_script QoScontroller" is an output line (just use to get a timestamp) to notify the start of the **commandRunner** running the **QoSscript.rb** that give the QoScorrection after receiving the ResourceUsage. 
It should totally give a JSON, i will try to rework it to return a JSON.